ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsFetch.scala

```diff
@@ -289,85 +291,112 @@ private class ContractsFetch(
+                stepsAndOffset.in <~ getActiveContracts(
+                  jwt,
+                  ledgerId,
+                  transactionFilter(parties, List(templateId)),
+                  true,
+                )(lc)
+                (stepsAndOffset.out0, stepsAndOffset.out1)
+
+              case (AbsoluteBookmark(_), _) | (LedgerBegin, true) =>
+                val stepsAndOffset = builder add transactionsFollowingBoundary(txnK)
+                stepsAndOffset.in <~ Source.single(startOffset)
+                (
+                  (stepsAndOffset: FanOutShape2[_, ContractStreamStep.LAV1, _]).out0,
+                  stepsAndOffset.out1,
+                )
+            }
 
-          val transactInsertsDeletes = Flow
-            .fromFunction(
-              jsonifyInsertDeleteStep(
-                (_: InsertDeleteStep[Any, lav1.event.CreatedEvent]),
-                templateId,
+            val transactInsertsDeletes = Flow
+              .fromFunction(
+                jsonifyInsertDeleteStep(
+                  (_: InsertDeleteStep[Any, lav1.event.CreatedEvent]),
+                  templateId,
+                )
               )
-            )
-            .via(if (sjd.q.queries.allowDamlTransactionBatching) conflation else Flow.apply)
-            .map(insertAndDelete)
+              .via(if (sjd.q.queries.allowDamlTransactionBatching) conflation else Flow.apply)
+              .map(insertAndDelete)
 
-          idses.map(_.toInsertDelete) ~> transactInsertsDeletes ~> acsSink
-          lastOff ~> offsetSink
+            idses.map(_.toInsertDelete) ~> transactInsertsDeletes ~> acsSink
+            lastOff ~> offsetSink
 
-          ClosedShape
-        }
-      )
+            ClosedShape
+          }
+        )
 
-      val (acsQueue, lastOffsetFuture) = graph.run()
+        val (acsQueue, lastOffsetFuture) = graph.run()
+
+        for {
+          _ <- sinkCioSequence_(acsQueue)
+          offset0 <- connectionIOFuture(lastOffsetFuture)
+          offsetOrError <- offset0 max AbsoluteBookmark(absEnd.toDomain) match {
+            case ab @ AbsoluteBookmark(newOffset) =>
+              ContractDao
+                .updateOffset(parties, templateId, newOffset, offsets)
+                .map(_ => ab)
+            case LedgerBegin =>
+              fconn.pure(LedgerBegin)
+          }
+        } yield offsetOrError
+      }
+  }
 
-      for {
-        _ <- sinkCioSequence_(acsQueue)
-        offset0 <- connectionIOFuture(lastOffsetFuture)
-        offsetOrError <- offset0 max AbsoluteBookmark(absEnd.toDomain) match {
-          case ab @ AbsoluteBookmark(newOffset) =>
-            ContractDao
-              .updateOffset(parties, templateId, newOffset, offsets)
-              .map(_ => ab)
-          case LedgerBegin =>
-            fconn.pure(LedgerBegin)
+  private def debugLogAction[T, C](
+      actionDescription: String
+  )(block: => T)(implicit lc: LoggingContextOf[C]): T = {
+    if (logger.debug.isEnabled) {
```

Comment:
Don't want to block this PR any further, but looking at this again, I don't think we need this branching. The argument to `logger.(debug|error|...)` is lazily evaluated, so if debug is not enabled, the message is never computed. We would be adding an additional `System.nanoTime()` call, but I'd expect that to be pretty cheap in comparison to the stuff we're doing in `block` and it may likely still be useful to know the duration in the case of an Exception

Reply:
as far as I know the logger.(debug|error|...) are Vals that are initialised by the Contextualization Logger and that is not lazy, each level is an object itself ... well maybe it change since I worked on logs long time ago..... I will take your word on this